zbus_macros/src/iface.rs

```diff
@@ -527,37 +572,75 @@ pub fn expand(args: AttributeArgs, mut input: ItemImpl) -> syn::Result<TokenStre
                         quote!(self.#ident()#method_await)
                     };
 
-                    let prop_changed_method = quote!(
-                        pub async fn #prop_changed_method_name(
-                            &self,
-                            signal_context: &#zbus::object_server::SignalContext<'_>,
-                        ) -> #zbus::Result<()> {
-                            let mut changed = ::std::collections::HashMap::new();
-                            let value = <#zbus::zvariant::Value as ::std::convert::From<_>>::from(#prop_value_handled);
-                            changed.insert(#member_name, &value);
-                            #zbus::fdo::Properties::properties_changed(
-                                signal_context,
-                                #zbus::names::InterfaceName::from_static_str_unchecked(#iface_name),
-                                &changed,
-                                &[],
-                            ).await
+                    let properties_changed_call = match p.emits_changed_signal {
+                        PropertyEmitsChangedSignal::True => {
+                            quote!(
+                                let mut changed = ::std::collections::HashMap::new();
+                                let value = <#zbus::zvariant::Value as ::std::convert::From<_>>::from(#prop_value_handled);
+                                changed.insert(#member_name, &value);
+                                #zbus::fdo::Properties::properties_changed(
+                                    signal_context,
+                                    #zbus::names::InterfaceName::from_static_str_unchecked(#iface_name),
+                                    &changed,
+                                    &[],
+                                ).await
+                            )
                         }
-                    );
+                        PropertyEmitsChangedSignal::Invalidates => {
+                            quote!(
+                                #zbus::fdo::Properties::properties_changed(
+                                    signal_context,
+                                    #zbus::names::InterfaceName::from_static_str_unchecked(#iface_name),
+                                    &::std::collections::HashMap::new(),
+                                    &[],
```

Comment:
Didn't catch this the last time. The last argument here is supposed to be the list of all invalidated properties. So it think this should be `&[#member_name)],`. It would be great to have a test case for this. Perhaps add the case to both interface and proxy side in `zbus/tests/e2e.rs`?

Reply:
This code goes in the `properties_changed_method`, so I do not think that we should be reporting that the property is getting invalidated when we call the `changed` method. However, the `prop_invalidate_method` code should have `&[#member_name]`, this is a few lines below this comment

Reply:
> This code goes in the `properties_changed_method`, so I do not think that we should be reporting that the property is getting invalidated when we call the `changed` method. However

Not sure I follow. This is as per the behavior defined by [the spec](url):

```
If one or more properties change on an object, the org.freedesktop.DBus.Properties.PropertiesChanged signal may be emitted (this signal was added in 0.14):

              org.freedesktop.DBus.Properties.PropertiesChanged (STRING interface_name,
                                                                 ARRAY of DICT_ENTRY<STRING,VARIANT> changed_properties,
                                                                 ARRAY<STRING> invalidated_properties);
        

where changed_properties is a dictionary containing the changed properties with the new values and invalidated_properties is an array of properties that changed but the value is not conveyed. 
```

Otherwise, the client gets a signal w/o any information at all and hence not useful.

> However, the `prop_invalidate_method` code should have `&[#member_name]`, this is a few lines below this comment

Good that you mentioned. I also missed that you are adding a new separate method for invalidating. I think this is not needed. The method should still be `_changed`. The behavior is then defined by the type of property declared by the user.